pennylane_lightning_gpu/src/algorithms/AdjointDataGPU.hpp

```diff
@@ -0,0 +1,252 @@
+     */
+    [[nodiscard]] auto getObsWires() const
+        -> const std::vector<std::vector<size_t>> & {
+        return obs_wires_;
+    }
+
+  private:
+    const std::vector<std::string> obs_name_;
+    const std::vector<param_var_t> obs_params_;
+    const std::vector<std::vector<size_t>> obs_wires_;
+};
+
+template <class T> class cuOpsData {
+  private:
+    size_t num_par_ops_;
+    size_t num_nonpar_ops_;
+    const std::vector<std::string> ops_name_;
+    const std::vector<std::vector<T>> ops_params_;
+    const std::vector<std::vector<std::string>> ops_hyperparams_;
+    const std::vector<std::vector<size_t>> ops_wires_;
+    const std::vector<bool> ops_inverses_;
+    const std::vector<std::vector<std::complex<T>>> ops_matrices_;
+
+  public:
+    /**
+     * @brief Construct an cuOpsData object, representing the serialized
+     * operations to apply upon the `%StateVector`.
+     *
+     * @param ops_name Name of each operation to apply.
+     * @param ops_params Parameters for a given operation ({} if optional).
+     * @param ops_hyperparams Hyper-parameters for a given observable operation
+     *  in string ({} if None).
+     * @param ops_wires Wires upon which to apply operation
+     * @param ops_inverses Value to represent whether given operation is
+     * adjoint.
+     * @param ops_matrices Numerical representation of given matrix if not
+     * supported.
+     */
+    cuOpsData(const std::vector<std::string> ops_name,
+              const std::vector<std::vector<T>> ops_params,
+              const std::vector<std::vector<std::string>> ops_hyperparams,
+              const std::vector<std::vector<size_t>> ops_wires,
+              const std::vector<bool> ops_inverses,
+              const std::vector<std::vector<std::complex<T>>> ops_matrices)
+        : ops_name_{std::move(ops_name)}, ops_params_{std::move(ops_params)},
+          ops_hyperparams_{std::move(ops_hyperparams)}, ops_wires_{std::move(
+                                                            ops_wires)},
+          ops_inverses_{std::move(ops_inverses)}, ops_matrices_{
+                                                      std::move(ops_matrices)} {
+        num_par_ops_ = 0;
+        for (const auto &p : ops_params) {
+            if (!p.empty()) {
+                num_par_ops_++;
+            }
+        }
+        num_nonpar_ops_ = ops_params.size() - num_par_ops_;
+    };
+
+    /**
+     * @brief Construct an cuOpsData object, representing the serialized
+     operations to apply upon the `%StateVector`.
+     *
+     * @see  cuOpsData(const std::vector<std::string> &ops_name,
+            const std::vector<std::vector<T>> &ops_params,
+            const std::vector<std::vector<std::string>> ops_hyperparams,
+            const std::vector<std::vector<size_t>> &ops_wires,
+            const std::vector<bool> &ops_inverses,
+            const std::vector<std::vector<std::complex<T>>> &ops_matrices)
+     */
```

Comment:
Could you please briefly speak about the difference between this and the full constructor? 
I think this could improve the documentation.

Reply:
Ah! This constructor doesn't require operations' matrices as input which is being called for supported operations. I thought it's clear... thanks! I added a note there. Let me know if you still think that I should be more descriptive.